scipy/linalg/tests/test_lapack.py

```diff
@@ -1645,3 +1645,134 @@ def test_getc2_gesc2():
         else:
             assert_array_almost_equal(desired_cplx.astype(dtype),
                                       x/scale, decimal=4)
+
+
+@pytest.mark.parametrize("dtype", DTYPES)
+@pytest.mark.filterwarnings("ignore:Casting complex")
+def test_gttrf_gttrs(dtype):
+    # The test uses ?gttrf and ?gttrs to solve a random system for each dtype,
+    # tests that the output of ?gttrf define LU matricies, that input
+    # parameters are unmodified, transposal options function correctly, that
+    # incompatible matrix shapes raise an error, and singular matrices return
+    # non zero info.
+
+    np.random.seed(42)
+    n = 10
+    rtol = 250*np.finfo(dtype).eps  # set test tolerance appropriate for dtype
+
+    # create the matrix in accordance with the data type
+    du = my_rand(n-1, dtype=dtype)
+    d = my_rand(n, dtype=dtype)
+    dl = my_rand(n-1, dtype=dtype)
+
+    diag_cpy = [dl.copy(), d.copy(), du.copy()]
+
+    A = np.diag(d) + np.diag(dl, -1) + np.diag(du, 1)
+    x = np.random.rand(n)
+    b = A @ x
+
+    gttrf, gttrs = get_lapack_funcs(('gttrf', 'gttrs'), dtype=dtype)
+
+    _dl, _d, _du, du2, ipiv, info = gttrf(dl, d, du)
+    # test to assure that the inputs of ?gttrf are unmodified
+    assert_array_equal(dl, diag_cpy[0])
+    assert_array_equal(d, diag_cpy[1])
+    assert_array_equal(du, diag_cpy[2])
+
+    # generate L and U factors from ?gttrf return values
+    # L/U are lower/upper triangular by construction (initially and at end)
+    U = np.diag(_d, 0) + np.diag(_du, 1) + np.diag(du2, 2)
+    L = np.eye(n, dtype=dtype)
+
+    for i, m in enumerate(_dl):
+        # L is given in a factored form.
+        # See http://www.hpcavf.uclan.ac.uk/softwaredoc/sgi_scsl_html/sgi_html/ch03.html
+        piv = ipiv[i] - 1
+        # right multiply by permutation matrix
+        L[:, [i, piv]] = L[:, [piv, i]]
+        # right multiply by Li, rank-one modification of identity
+        L[:, i] += L[:, i+1]*m
+
+    # one last permutation
+    i, piv = -1, ipiv[-1] - 1
+    # right multiply by final permutation matrix
+    L[:, [i, piv]] = L[:, [piv, i]]
+
+    _A = L @ U
+    # check that the outputs of ?gttrf define an LU decomposition of A
+    assert_allclose(A, _A, atol=rtol)
```

Comment:
If we don't use `_A` anywhere else I think explicitly performing the LU (re?)-decomposition inside the assert will be just as clear eg:

```python
# check that the outputs of ?gttrf define an LU decomposition of A
assert_allclose(A,  L @ U, atol=rtol)
```

second the relative tolerance is used to set the absolute tolerance in ``assert_allclose(..., atol=rtol)`` is that meant to be rtol?

Reply:
Nice catch with the atol=rtol–due to some problems I discovered with using relative tolerance when comparing zeros, I think that atol does need to be changed to something here.  I added in a dtype dependent atol at the top for this.